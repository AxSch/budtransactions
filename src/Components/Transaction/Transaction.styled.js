import styled from 'styled-components'

export const StyledTransaction = styled.div`
    display: flex;
    padding-top: 10px;
    background-color: ${({ isToBeRemoved }) => isToBeRemoved ? 'rgb(255, 122, 127)': 'rgb(255,255,255)'};
    position: relative;
    z-index: 2;
`

export const StyledTransactionDetails = styled.div`
    display: flex;
    width: 100%;
    flex-direction: column;
    justify-content: center;
    border-bottom: 0.5px solid rgb(233, 233, 233);
`

export const StyledTransactionIcon = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    background-color: none;
    padding: 20px;
    > img {
        border: 1px solid rgb(228, 233, 234);
        border-radius: 50%;
        height: 42px;
        width: 42px;
    }
`

export const StyledTransactionAmount = styled.div`
    display: flex;
    flex-direction: column;
    font-weight: 600;
    justify-content: center;
    padding: 26px;
    border-bottom: 0.5px solid rgb(233, 233, 233);
    color: ${({ isToBeRemoved }) => isToBeRemoved ? 'rgb(205, 38, 49)': 'rgb(0, 0, 0)'};
`

export const StyledTransactionDescription = styled.div`
    font-weight: 500;
    display: flex;
    font-size: 18px;
    padding: 10px 0px;
    color: ${({ isToBeRemoved }) => isToBeRemoved ? 'rgb(205, 38, 49)': 'rgb(0, 0, 0)'};
`

export const StyledTransactionCategory = styled.div`
    display: flex;
    font-size: 14px;
    color: rgb(134, 134, 136);
    margin-bottom: 10px;
    color: ${({ isToBeRemoved }) => isToBeRemoved ? 'rgb(247, 94, 101)': 'rgb(0, 0, 0)'};
`

export const Overlay = styled.div`
    background-color: ${({ isToBeRemoved }) => isToBeRemoved ? 'rgba(255, 122, 127, 0.75)': 'none'};
    position: absolute;
    top: 0;
    bottom: 0;
    right: 0;
    left: 0;
    z-index: ${({ isToBeRemoved }) => isToBeRemoved ? 3: -1};
    border-bottom: ${({ isToBeRemoved }) => isToBeRemoved ? '0.5px solid rgb(233, 233, 233)': 'rgb(255,255,255)'};
`
