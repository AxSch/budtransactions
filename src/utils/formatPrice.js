import currency from 'currency.js'

const gbp = value => currency(value, { symbol: 'Â£' })

export const formatPrice = (priceDetails) => {
    return gbp(priceDetails.value).format()
}
