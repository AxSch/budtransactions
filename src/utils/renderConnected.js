import React from 'react';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import transactionsSlice from '../reducers/transactions/transactionsSlice';

const renderConnected = (
  ui, {
    store = createStore(transactionsSlice, {}),
    ...renderOptions
  } = {}
) => {
  const Wrapper = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );
  return render(ui, { wrapper: Wrapper, ...renderOptions});
};

export default renderConnected;

