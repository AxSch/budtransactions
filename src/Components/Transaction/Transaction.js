import React, { useEffect, useState } from 'react'
import {
    StyledTransaction,
    StyledTransactionIcon,
    StyledTransactionDetails,
    StyledTransactionDescription,
    StyledTransactionCategory,
    StyledTransactionAmount,
    Overlay
} from './Transaction.styled'
import { markedForRemoval, selectIsEdit } from '../../reducers/transactions/transactionsSlice'
import { useDispatch, useSelector } from 'react-redux'
import { formatPrice } from '../../utils/formatPrice'


const Transaction = ({ transaction }) => {
    const [isToBeRemoved, setToBeRemoved] = useState(false)
    const isEdit = useSelector(selectIsEdit)
    const dispatch = useDispatch()

    useEffect(() => {
        if(!isEdit) {
            setToBeRemoved(false)
        }
    }, [isEdit])
    
    const handleRemoval = (transactionId) => {
        if (isEdit) {
            setToBeRemoved(!isToBeRemoved)
            dispatch(markedForRemoval({ id: transactionId, toBeRemoved: !isToBeRemoved }))
        }
    }

    return (
        <>
            <StyledTransaction data-testid="transaction" onClick={() => handleRemoval(transaction.id)} isEdit={isEdit}>
                <Overlay isToBeRemoved={isToBeRemoved} />
                <StyledTransactionIcon data-testid="transaction-bank-icon">
                    <img alt="bank-icon" src={transaction.product.icon} />
                </StyledTransactionIcon>
                <StyledTransactionDetails>
                    <StyledTransactionDescription data-testid="transaction-description" isToBeRemoved={isToBeRemoved}>
                        {transaction.description}
                    </StyledTransactionDescription>
                    <StyledTransactionCategory data-testid="transaction-category" isToBeRemoved={isToBeRemoved}>
                        <span>
                            {transaction.category}
                        </span>
                    </StyledTransactionCategory>
                </StyledTransactionDetails>
                <StyledTransactionAmount data-testid="transaction-amount" isToBeRemoved={isToBeRemoved}>
                    {formatPrice(transaction.amount)}
                </StyledTransactionAmount>
            </StyledTransaction>
        </>
    )
}

export default Transaction
