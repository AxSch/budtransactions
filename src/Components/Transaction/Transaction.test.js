import { fireEvent } from '@testing-library/react';
import * as ReactRedux from 'react-redux';
import React from 'react'
import Transaction from './Transaction'

import renderConnected from '../../utils/renderConnected'
import { formatPrice } from '../../utils/formatPrice';

const initialState = {
        isEdit: true,
        userTransactions: {},
        transactionsToRemove: []
}

const mockTransaction = {
    id: "13acb877dc4d8030c5dacbde41235433d3496a2ae3c000db4c793bda9c3228baca1a28",
    date: "2018-03-05",
    description: "Five guys",
    category: "Food / Drink",
    currency: "GBP",
    amount: {
        value: 32,
        currency_iso: "GBP"
    },
    product: {
        id: 4279,
        title: "Lloyds Bank",
        icon: "https://storage.googleapis.com/budcraftstorage/uploads/products/lloyds-bank/Llyods_Favicon-1_161201_091641.jpg"
    }
}
let getByTestId, container

describe('should be display transactions correctly to the user', () => {
    const mockSelector = jest.fn();
    const mockDispatch = jest.fn();
    ReactRedux.useSelector = jest.fn().mockImplementation(() => mockSelector)
    ReactRedux.useDispatch = jest.fn().mockImplementation(() => mockDispatch)
    
    beforeEach(() => {
        const utils = renderConnected(<Transaction transaction={mockTransaction} />, { initialState });
        getByTestId = utils.getByTestId;
    });

    afterEach(() => {
        jest.clearAllMocks()
    })

    it('renders successfully', () => {
        expect(getByTestId('transaction')).toBeInTheDocument()
        expect(container).toMatchSnapshot();
    });

    it('should render the necessary transaction information', () => {
        const transactionIcon = getByTestId('transaction-bank-icon')
        const transactionDesc = getByTestId('transaction-description')
        const transactionCategory = getByTestId('transaction-category')
        const transactionAmount = getByTestId('transaction-amount')
        const mockFormattedPrice = formatPrice(mockTransaction.amount)
        
        expect(transactionIcon).toBeInTheDocument()
        expect(transactionDesc.textContent).toBe(mockTransaction.description)
        expect(transactionCategory.textContent).toBe(mockTransaction.category)
        expect(transactionAmount.textContent).toBe(mockFormattedPrice)
    });

});