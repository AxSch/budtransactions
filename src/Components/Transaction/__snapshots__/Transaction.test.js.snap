// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be display transactions correctly to the user renders successfully 1`] = `undefined`;
